검색할 장소 검색

randychoi95 opened the pull request:
# 검색할 장소 검색 구현
- 지도에서 자동 검색 기능 구현
- 검색된 장소 표출
- 닫기 버튼 누르시 PlaceSearcher Riblet detach 및 라우팅(dismiss) 추가
- 검색 후 해당 장소로 이동

<img src = "https://github.com/f-lab-edu/PlaceStory/assets/60861313/cc722ae0-caa6-4126-80e5-dd876b585cd8" width="300">

Commit:
:lipstick: 검색할 장소 검색 UI 구현

- 지도에서 자동 검색 기능 구현
- 검색된 장소 표출
- 닫기 버튼 누르시 PlaceSearcher Riblet detach 및 라우팅(dismiss) 추가

f-lab-max commented on PlaceStory/MyLocation/Sources/PlaceSearcher/Cell/SearchTableViewCell.swift:
```diff
@@ -0,0 +1,59 @@
+    
+    // MARK: - UI Components
+    
+    lazy var placeLabel: UILabel = {
```
클로져에서 self를 쓰고 있지 않으니 lazy var 대신 let으로 대체할 수 있습니다.

randychoi95 replied:
let으로 변경하겠습니다.

f-lab-max commented on PlaceStory/MyLocation/Sources/PlaceSearcher/Cell/SearchTableViewCell.swift:
```diff
@@ -0,0 +1,59 @@
+import Utils
+import UIKit
+
+class SearchTableViewCell: UITableViewCell {
```
상속할 계획이 아닌 클래스는 final을 써주시면 성능에 도움이 됩니다.

f-lab-max commented on PlaceStory/MyLocation/Sources/PlaceSearcher/PlaceSearcherInteractor.swift:
```diff
@@ -17,7 +19,7 @@ protocol PlaceSearcherPresentable: Presentable {
 
 public protocol PlaceSearcherListener: AnyObject {
-    // TODO: Declare methods the interactor can invoke to communicate with other RIBs.
+    func placeSearcherDidTapClose()
```
네이밍 좋네요 (제가 쓰는 컨벤션과 같아서 👍 ㅋㅋ)

f-lab-max commented on PlaceStory/MyLocation/Sources/PlaceSearcher/PlaceSearcherInteractor.swift:
```diff
@@ -41,4 +43,25 @@ final class PlaceSearcherInteractor: PresentableInteractor<PlaceSearcherPresenta
+    }
+    
+    func didSelected(for result: MKLocalSearchCompletion) {
+        let searchReqeust = MKLocalSearch.Request(completion: result)
```
MapKit을 인터랙터가 직접 쓰면 인터랙터와 맵킷이 강하게 결합되어 테스트도 어렵고 느슨한 결합의 장점을 여러 누릴수가 없습니다. "장소 검색"은 변동적 의존성(volatile dependency)이므로 이것 또한 추상화하여 사용하는게 좋습니다.

f-lab-max commented on PlaceStory/MyLocation/Sources/PlaceSearcher/PlaceSearcherViewController.swift:
```diff
@@ -5,26 +5,235 @@
-    // TODO: Declare properties and methods that the view controller can invoke to perform
-    // business logic, such as signIn(). This protocol is implemented by the corresponding
-    // interactor class.
+    func didTappedCloseButton()
+    func didSelected(for result: MKLocalSearchCompletion)
```
didTap, didSelect 로 쓰는 것이 문법에 맞는 방법입니다. did 다음에는 동사원형을 씁니다.

f-lab-max commented on PlaceStory/MyLocation/Sources/PlaceSearcher/PlaceSearcherViewController.swift:
```diff
@@ -5,26 +5,235 @@
+    // MARK: - Property
+    
     weak var listener: PlaceSearcherPresentableListener?
+    private var searchCompleter = MKLocalSearchCompleter()
```
마찬가지로, 인터랙터와 뷰가 MapKit에 의존하지 않도록, 도메인와 플랫폼에 관련 기능을 넣고 MapKit과의 직접적인 의존성을 끊어주세요

randychoi95 replied:
조언해주신대로 구현해보겠습니다

Commit:
:sparkles: 장소 검색 후 이동

- 장소 검색후 위치로 이동
- 장소에 pin 등록

Commit:
:recycle: 리뷰 반영

- MapKit 관련 기능들 Domain, Platform으로 분산하여 interactor와 결합 줄이기
- lazy var -> let 으로 수정
- tableview cell final 키워드 추가

f-lab-max commented on PlaceStory/Domain/Sources/Repositories/MapServiceRepository.swift:
```diff
@@ -0,0 +1,15 @@
+import MapKit
+
+public protocol MapServiceRepository {
+    func searchPlace(from text: String) -> AnyPublisher<[MKLocalSearchCompletion], Never>
```
MapKit 모델 MKLocalSearchCompletion을 외부로 유출하고 있기 때문에 여전히 인터랙터가 import MapKit을 해주고 있죠. UI 계층과 MapKit과의 의존성이 여전히 깨지지 않은 상태입니다. 가령, MapKit을 다른 라이브러리로 교체하기로 한다면 UI까지 코드를 수정해줘야하는 상황인거죠. UI 레이어는 도메인에만 의존하도록 수정해주세요.

f-lab-max reviewed (changes requested):
PR에 2가지 작업이 포함되고 변경을 하면서 점점 커지고 있습니다. 현재 500줄 이상. 작업 단위를 쪼개서 두개의 PR로 만들어보세요

Commit:
:refactor: 리뷰 반영

- Domain layer에서 지도 검색/검색 장소에 대한 Entity를 만듦
- interactor에서 Entity를 사용하여 MapKit과 분리

Commit:
:white_check_mark: Add MyLocation Riblet's Interactor Unit Test

- MyLocation Riblet Mock 객체들 생성
- interactor에 대한 test case 추가

Commit:
:white_check_mark: Update MyLocation Riblet's Interactor Unit Test and Add Router Unit Test

- 라우터 확인을 위한 PlaceSearch Mock 객체들 구현
- MyLocation Builder 타입 protocol 타입으로 변환

randychoi95 committed:
:refactor: 리뷰 반영 - MapKit 의존성 분리

- MKLocalSearchCompletion의 객체의 프로퍼티들을 받는 도메인 모델을 구현하여 도메인 모델을 사용하게 수정

randychoi95 committed:
Merge pull request #13 from f-lab-edu/feature/test-MyLocation

MyLocation Riblet Unit test 추가

randychoi95 committed:
Merge pull request #12 from f-lab-edu/feature/modify-MapService

Fix. 검색할 장소 검색 (#11)
